src/java/org/apache/cassandra/db/compaction/CompactionTask.java

```diff
@@ -209,7 +209,7 @@ private CompactionOperation createCompactionOperation(CompactionController contr
         {
             // The set of sstables has changed (one or more were excluded due to limited available disk space).
             // We need to recompute the overlaps between sstables.
-            controller.maybeRefreshOverlaps();
+            controller.refreshOverlaps();
```

Comment:
On first glance this does not appear sufficient to solve the CASSANDRA-18507 problem: the overlap tracker is built for the given set of sstables and refreshing does not change that set (i.e. it will still exclude the sstables in the original set, making it possible to miss data).

It looks like it works because the reference the tracker keeps is to the live originals set in the transaction (which is what `buildCompactionCandidates...` modifies); the compactions controller also seems to rely on its reference giving the up-to-date list. Perhaps add some comments to state that we are relying on the live view?

Reply:
Ok, I see what you mean - the compaction controller and tracker are both created with a set of sstables that, at least for them, do not change as they are iterators of an underlying collection. However `buildCompactionCandidates` does modify the underlying collection via the transaction, so refreshing the overlaps does result in a change because of that side effect change.

The existing comment here could be more explicit, maybe like:
```
            // The set of sstables has changed (one or more were excluded due to limited available disk space).
            // We need to recompute the overlaps between sstables. The iterators used in the compaction controller 
            // and tracker will reflect this changed set of sstables made in the `LifecycleTransaction` transaction,
            // so refreshing the overlaps will be based on the updated set of sstables.
```

Do you think that helps make it more clear? Happy to take suggestions on wording or where to place additional explanatory comments.

Reply:
Yes, this is definitely better and is probably sufficient because this test will fail immediately if someone changes the way the iterable is passed.

Optionally, we can also add something at [the location the controller is constructed](https://github.com/datastax/cassandra/blob/5fe9c507ae97ffdbee204e6adeff511ccf72f5f7/src/java/org/apache/cassandra/db/compaction/CompactionTask.java#L196).